rpc/metadata.go

```diff
@@ -0,0 +1,34 @@
+package rpc
+
+import (
+	"context"
+	"net/http"
+)
+
+type mdHeaderKey struct{}
+
+// NewContextWithHeaders is used to attach http headers into the context
+func NewContextWithHeaders(ctx context.Context, header http.Header) context.Context {
+	return context.WithValue(ctx, mdHeaderKey{}, header)
+}
+
+// HeadersFromContext is used to extract http headers from the context
+func HeadersFromContext(ctx context.Context) http.Header {
+	value := ctx.Value(mdHeaderKey{})
+	if value == nil {
+		return nil
+	}
+	return value.(http.Header)
+}
+
+// addHeadersFromContext is used to extract http headers from the context and inject it into the provided headers
+func addHeadersFromContext(ctx context.Context, headers http.Header) {
+	if kvs := HeadersFromContext(ctx); kvs != nil {
```

Comment:
See https://go.dev/play/p/LKrW4z8vMxE 

You can omit this outer clause, and go directly to the `for .. range kvs`